### PR TITLE
add: [plugins] hashlookup plugin

### PR DESCRIPTION
**Pull request by gallypette**
Hi there, here is the hashlookup plugin come back from the dead after the forced push episode. From a helicopter view, it may make sense to remove everything related to SQL and keep it as simple as possible.

It does not work anymore and I don't really get why:

graphoscope-hashlookup

### CODE REVIEW

File: plugins/src/hashlookup/hashlookup.go [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+	"regexp"
+	"strings"
+	"sync"
+)
+
+/*
+ * Check "pdk/plugin.go" for the built-in plugin functions description
+ */
+
+func (p *plugin) Source() *pdk.Source {
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]: Rename method to `Conf()`

File: plugins/src/hashlookup/hashlookup.go [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+
+		for _, types := range relation.To.VarTypes {
+			types.RegexCompiled = regexp.MustCompile(types.Regex)
+		}
+	}
+
+	fmt.Printf("HTTP %s: %#v\n\n", source.Name, p)
+	return nil
+}
+
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go:62]:
```go
// New method for Web GUI field names autocomplete:
func (p *plugin) Fields() ([]string, error) {
        return p.source.QueryFields, nil
}
```

File: plugins/src/hashlookup/hashlookup.go [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+func (p *plugin) Search(stmt *sqlparser.Select) ([]map[string]interface{}, map[string]interface{}, error) {
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]: Current list of returned types:
```
([]map[string]interface{}, map[string]interface{}, map[string]interface{}, error)
```

File: plugins/src/hashlookup/hashlookup.go [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+
+	// Storage for the results to return
+	results := []map[string]interface{}{}
+
+	searchFields, err := p.convert(stmt)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var bodies []*bytes.Buffer
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]: `debug` info was recently added as returned value. Will also be visible in browser console if enabled
```go
searchFields, err := p.convert(stmt)
if err != nil {
	return nil, nil, nil, err
}

// Debug info
debug := make(map[string]interface{})
var bodies []*bytes.Buffer
```

File: plugins/src/hashlookup/hashlookup.go [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+
+	/*
+	 * Send indicators to get results back
+	 */
+	bodies, err = p.request(searchFields)
+	var unpacked []string
+	if err != nil {
+		return nil, nil, err
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return nil, nil, debug, err
```

File: plugins/src/hashlookup/hashlookup.go [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+
+	/*
+	 * Receive hits and deserialize them
+	 */
+
+	// Attempt to unpack arrays if we get results from the bulk endpoint
+	for _, body := range bodies {
+		jsonParsed, err := gabs.ParseJSONBuffer(body)
+		if err != nil {
+			return nil, nil, err
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return nil, nil, debug, err
```

File: plugins/src/hashlookup/hashlookup.go [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+		// when the amount of returned entries is too large
+		stats := pdk.NewStats()
+
+		for _, field := range p.source.StatsFields {
+			stats.Fields[field] = sortedmap.New(10, desc.Int)
+		}
+
+		jsonParsed, err := gabs.ParseJSON([]byte(body))
+		if err != nil {
+			return nil, nil, err
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return nil, nil, debug, err
```

File: plugins/src/hashlookup/hashlookup.go [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+		children := new(gabs.Container)
+		parents := new(gabs.Container)
+
+		// TODO fetch children / parents according to the LIMIT
+		// if the children/parent key exist and the count hits hashlookup limits,
+		// we query hashlookup again up the the query' LIMIT
+
+		if jsonParsed.Exists("children") {
+			if err != nil {
+				return nil, nil, err
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return nil, nil, debug, err
```

File: plugins/src/hashlookup/hashlookup.go [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+			}
+			// Here we run the new query
+			gabsChildren, err := p.getRelated("children", jsonParsed.S("SHA-1").Data().(string), p.limit)
+			if err != nil {
+				return nil, nil, err
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return nil, nil, debug, err
```

File: plugins/src/hashlookup/hashlookup.go [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+				entryObj.ArrayAppend(jsonObj, "entries")
+			}
+			// Remove the children for the main received object
+			jsonParsed.Delete("children")
+		}
+		if jsonParsed.Exists("parents") {
+			// Here we run the new query
+			gabsParents, err := p.getRelated("parents", jsonParsed.S("SHA-1").Data().(string), p.limit)
+			if err != nil {
+				return nil, nil, err
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return nil, nil, debug, err
```

File: plugins/src/hashlookup/hashlookup.go [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+			}
+			parents = gabsParents.Path("parents")
+			if err != nil {
+				return nil, nil, err
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return nil, nil, debug, err
```

File: plugins/src/hashlookup/hashlookup.go [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+		}
+
+		entryObj.ArrayAppend(jsonParsed, "entries")
+
+		finalJSONParsed.Merge(entryObj)
+	}
+
+	err = json.NewDecoder(strings.NewReader(finalJSONParsed.S("entries").String())).Decode(&entries)
+	if err != nil {
+		return nil, nil, err
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return nil, nil, debug, err
```

File: plugins/src/hashlookup/hashlookup.go [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+	for _, entry := range entries {
+
+		// Stop when results count is over the limit
+		if counter >= p.limit {
+			// Uncomment in real plugin
+			//cancel()
+
+			top, err := stats.ToJSON(p.source.Name)
+			if err != nil {
+				return nil, nil, err
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return nil, nil, debug, err
```

File: plugins/src/hashlookup/hashlookup.go [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+			}
+
+			return nil, top, nil
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return nil, top, debug, nil
```

File: plugins/src/hashlookup/hashlookup.go [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+				 * Add current entry to the list to return
+				 */
+				mx.Lock()
+				results = append(results, result)
+				mx.Unlock()
+			}
+		}
+	}
+
+	return results, nil, nil
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]:
```go
return results, nil, debug, nil
```

File: plugins/src/hashlookup/hashlookup.go (hunk 2 above)
Comment (aleksejsv) [reply plugins/src/hashlookup/hashlookup.go:62]: ..this new method can be simply copy-pasted, if there is no way to get automatically a list of available field names to query

**Comment by aleksejsv** [comment]
`SourcePlugin` interface's methods, which must be implemented have changed a little. I've added comments for needed changes. With up-to-date methods plugin is starting successfully.

Check https://github.com/cert-lv/graphoscope/blob/master/pdk/plugin.go for a list of methods and their return values to be implemented.

`SourcePlugin` - for plugins for data sources (your case);
`ProcessorPlugin` - for plugins to process/enrich data before sending to the requester

**Comment by gallypette** [comment]
Hi, it compiles and it works.

**Comment by gallypette** [comment]
thanks.

**Pull request merged**